### PR TITLE
[Work in progress] Upgrade to Node 6

### PR DESCRIPTION
**Pull request by benjamn**
This pull request is intended to accumulate the necessary changes to make Meteor fully compatible with Node 6.

Fortunately, many of the changes we made to support Node 4 should also ease the transition to Node 6, but some blockers remain:

- [x] npm package compatibility, e.g. `sqlite3` (possibly solved by upgrading to `sqlite3@3.1.8`)
- [x] V8 memory limits (possibly solved by `--max-old-space-size=8192`)
- [x] excessive memory usage while building (unsolved)
- [x] high (~100%) CPU usage while app is running idle

In order to use this experimental version of Meteor, check out the [`wip-upgrade-to-node-6` branch](https://github.com/meteor/meteor/tree/wip-upgrade-to-node-6), regenerate the dev bundle, and run the `meteor` script from a checkout:

```sh
cd path/to/meteor
git fetch origin
git checkout -t origin/wip-upgrade-to-node-6
git submodule update --init --recursive
scripts/generate-dev-bundle.sh
./meteor --help
```

See [these instructions](https://github.com/meteor/meteor/blob/devel/Development.md#running-from-a-git-checkout) for help with running Meteor from a checkout.

### CODE REVIEW

**Comment by benjamn** [comment]
Node 6.0.0 was only released yesterday, so naturally there are incompatibilities with some npm packages used by Meteor.

The most immediate blocker appears to be `sqlite3`: https://github.com/mapbox/node-sqlite3/issues/634

**Comment by trajano** [comment]
Can the target milestone for this be something other than 1.4? It should really be after Node 4 support is added right?

**Comment by tcastelli** [comment]
As I understand, this doesn't go after v4. We have to choose if we start with 4 or we start directly with 6. IMO we should try to start with 6, because since there are going to be problems anyways, better them be in favor of becoming integrated with the latest LTS version.

File: scripts/generate-dev-bundle.ps1 [hunk 1]
```diff
@@ -100,7 +99,6 @@ mkdir -Force "${DIR}\b\p"
 cd "${DIR}\b\p"
 node "${CHECKOUT_DIR}\scripts\dev-bundle-tool-package.js" | Out-File -FilePath package.json -Encoding ascii
 npm install
-npm dedupe
```
Comment (trajano) [inline scripts/generate-dev-bundle.ps1]: What does it output when it "fails"?

**Comment by suhail-ansari-apconic** [comment]
Why not upgrading node to v6 be a intermediate release instead of 1.4. That way the whole system can be stabilized before meteor is broken into different npm packages.

**Comment by benjamn** [comment]
Our current vision for Meteor 1.5 is that you will be able to use whatever (relatively recent) version of Node you like: https://github.com/meteor/meteor/pull/7072

In all likelihood, Meteor 1.5 will be released before Node v6 is ready for production use, so most Meteor developers will want to keep using Node v4 with Meteor 1.5 for a few months at least. When Node v6 is ready, the transition should be straightforward, though.

**Comment by jykae** [comment]
@benjamn that sounds great vision, I've been a bit worried about both MongoDB and Node.js version support.

**Comment by benjamn** [comment]
This pull request is pretty out of date now. I'll create a new PR when we're ready to jump from Node 4 to Node 6 (likely after Meteor 1.4). Closing in the meantime.

**Comment by Type1J** [comment]
The current Node LTS is v6.9.1 as of 2016-11-14. What is the new issue number to watch for Node 6 compatibility?

**Comment by HLFH** [comment]
Yes, it's a big issue. https://github.com/wekan/wekan is stuck to Node.js 0.10.40. And a lot of softwares are now using Node.js v6. So it would be great to have Meteor compatible with Node.js v6.

@Type1J You should create a Node 6 compatibility issue. There is none right now. Let us know, thanks.

**Comment by benjamn** [comment]
Let's just reopen this PR, shall we?

**Comment by benjamn** [comment]
@HLFH You (and/or whoever is working on Wekan) really should do everything you can to upgrade to Meteor 1.4(.2.1) and Node 4. Jumping directly to Node 6 will be very difficult if you're not already on Node 4.

**Comment by benjamn** [comment]
Whoa, `meteor test-packages` (the Travis CI suite) is actually passing! I must admit I expected basically none of the tests to pass. Progress!

**Comment by vlasky** [comment]
@benjamn does the passing of the tests mean that there will soon be a Node 6-enabled version of Meteor, or are there more hurdles to overcome?

**Comment by brucejo75** [comment]
Will [debugging with chrome tools](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27#.sam9dduaz) be supported just by upgrading to v6?

**Comment by Type1J** [comment]
@benjamn Does this PR's branch now have a working Meteor on Node 6? Are there any remaining issues?

**Comment by abernix** [comment]
@vlasky @Type1J Meteor with Node 6 is not ready _right now_ and there is still work to be done.  Per the [Roadmap.md](https://github.com/meteor/meteor/blob/devel/Roadmap.md) and the subject of this PR, it is a work in progress.  The comment above by @benjamn about the Travis CI suite working is only part of the tests.  As you can see by the CircleCI tests on this PR (a separate suite of tests entirely), they failed [spectacularly](https://circleci.com/gh/meteor/meteor/3068).  Not to say that this means there is _substantial_ work, but there are areas that need attention.

If you'd like to **try** using Meteor with Node 6 there are instructions in the initial post of this PR but I really wouldn't expect things to work flawlessly.  Your help testing would be appreciated, but if you choose to try it out, please embark on that adventure with a willingness to help report/debug/solve the problems you encounter.

**Comment by deadjdona** [comment]
Just checked on app: Meteor latest version with Node versions, used n to switch node versions.

meteor + node 7 = 100% CPU
meteor + node 6 = 100% CPU
meteor + node 4.7 - works fine.

**Comment by benjamn** [comment]
> Will [debugging with chrome tools](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27#.sam9dduaz) be supported just by upgrading to v6?

@brucejo75 That is certainly my hope!

**Comment by partounian** [comment]
@deadjdona Just to comment on that CPU usage. My runs at 100% for all node version usage, it only lowers when running meteor release 1.4.1.2 (will go as low as 15% during non-building).

**Comment by laosb** [comment]
It seems that we should consider Node 8.0 next - can't wait!

**Comment by brucejo75** [comment]
Hi @benjamn, I just entered #8590 in the hopes that moving to Node 6 may get this fixed?

**Comment by yorrd** [comment]
@benjamn I've noticed you bumped the version to 1.6. Does that mean we'll have to wait for like another few months? I was so happy to hear we'll get this for 1.5 (which hopefully arrives soon-ish?) o.o 

I've already collected a list of several important libraries (discord.js or quill.js for example) which would step up my meteor game by so much. Clients are asking for this stuff constantly...

**Comment by benjamn** [comment]
@yorrd That's right, this is far too big a change to include in Meteor 1.5 at this point. Getting Meteor 1.6 stable will be the top priority after Meteor 1.5 is done. There's a good chance we'll release 1.6 before 1.5.1, unless the remaining problems turn out to be really hard to solve, but I'm not too worried about that.

**Comment by yorrd** [comment]
@benjamn so, I know you don't want to give an ETA, but are we talking days, weeks, months?

**Comment by benjamn** [comment]
Well, would you look at that, the entire test suite is passing!

@yorrd I didn't want to make any promises before I saw that green checkmark, since it was anyone's guess how long that could have taken, but now that tests are passing, I anticipate roughly a month of remaining work.

Here's an incomplete list of things I know still need to be done:
- [ ] Audit the codebase to catch up with any breaking changes between Node 4 and 5: https://github.com/nodejs/node/wiki/Breaking-changes-between-v4-and-v5
- [ ] Audit the codebase to catch up with any breaking changes between Node 5 and 6: https://github.com/nodejs/node/wiki/Breaking-changes-between-v4-and-v5
- [ ] Reimplement `meteor debug` to use [Chrome DevTools](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27) instead of `node-inspector` (👋  @brucejo75), which is such a better debugging experience it's not even funny.
- [ ] Get `meteor-tool` to build on Linux 32bit (runs out of memory currently).
- [ ] Get `meteor-tool` to build on Windows, and consider adding a 64bit Windows architecture. Note: I think both of these tasks will be easy, but as of right now our Windows build farm doesn't meet the minimum OS requirements of Node 6.
- [ ] Consider removing or reducing the use of `Fiber`s in Meteor's command-line tool, though they certainly still work in Node 6, and we've managed to work around the high CPU problems caused by previous versions of the `meteor-promise` package (which was the key to getting the tests to pass today).
- [ ] Release Meteor 1.5 and merge those changes into this branch (easy).

I know I'm forgetting important items here, so please feel free to comment with anything else you can think of!

**Comment by laosb** [comment]
@benjamn Just corrected the link for breaking changes between Node 5 and 6. Thank you for your hardwork!

**Comment by leosco** [comment]
I'm guessing it's not easy to contribute to Meteor on Windows? I'm getting this message when I try to run `scripts/generate-dev-bundle.sh` in Git Bash: 
```
MINGW64 ~/Projects/meteor (wip-upgrade-to-node-6)
$ scripts/generate-dev-bundle.sh
This OS not yet supported
```

**Comment by benjamn** [comment]
@leosco Try `scripts/generate-dev-bundle.ps1` (the PowerShell script we use to generate the Windows dev bundle)?

**Comment by leosco** [comment]
@benjamn thanks for the tip! Looks like it's working :) Only hoop to jump through so far is installing 7-Zip. Script crashes if it can't find `C:\Program Files\7-Zip\7z.exe` lol

**Comment by AndreasGalster** [comment]
I saw there's a pre-build already for 1.6. With Node 6, will this already have a reduced ecmascript package? Specifically, we're trying to build a meteor-server only with graphql and as much native es6 as possible on the server would be nice for faster builds :).

**Comment by benjamn** [comment]
@AndreasGalster Reducing the set of Babel plugins needed on the server is definitely a goal of Meteor 1.6, but the `1.6-pre.1` release is currently just Meteor 1.5 + <minimal changes to get Node 6 to work>.

**Comment by benjamn** [comment]
Superseded by https://github.com/meteor/meteor/pull/8728.

**Pull request closed without merging**